File: class15/java/PES1501-FuncionarioHeranca.java
/*
 * PES1501 - Funcionário com Herança e Polimorfismo
 * Aula 15: Diagrama de Classes e Orientação a Objetos
 * Projeto e Engenharia de Software - Prof. Giovane Barcelos
 *
 * Demonstra herança, polimorfismo e encapsulamento com uma hierarquia
 * Funcionario → Desenvolvedor, Gerente.
 *
 * Compilar/executar:
 *   javac PES1501-FuncionarioHeranca.java && java FuncionarioHeranca
 */

import java.util.*;

abstract class Funcionario {
    private String nome;
    private double salarioBase;

    public Funcionario(String nome, double salarioBase) {
        this.nome = nome;
        this.salarioBase = salarioBase;
    }

    public String getNome() { return nome; }
    public double getSalarioBase() { return salarioBase; }

    public abstract double calcularSalario();

    @Override
    public String toString() {
        return String.format("%s[nome=%s, salário=R$ %.2f]",
                getClass().getSimpleName(), nome, calcularSalario());
    }
}

class Desenvolvedor extends Funcionario {
    private String linguagem;

    public Desenvolvedor(String nome, double salarioBase, String linguagem) {
        super(nome, salarioBase);
        this.linguagem = linguagem;
    }

    @Override
    public double calcularSalario() {
        return getSalarioBase() * 1.15;
    }
}

class Gerente extends Funcionario {
    private double bonus;

    public Gerente(String nome, double salarioBase, double bonus) {
        super(nome, salarioBase);
        this.bonus = bonus;
    }

    public Gerente(String nome, double salarioBase) {
        this(nome, salarioBase, 2000.0);
    }

    @Override
    public double calcularSalario() {
        return getSalarioBase() + bonus;
    }
}

class FuncionarioHeranca {

    public static void processarFolha(List<Funcionario> funcionarios) {
        System.out.printf("%n%-20s %-18s %15s%n", "Nome", "Cargo", "Salário Final");
        System.out.println("-".repeat(55));
        double total = 0;
        for (Funcionario f : funcionarios) {
            double sal = f.calcularSalario();
            total += sal;
            System.out.printf("%-20s %-18s R$ %12.2f%n",
                    f.getNome(), f.getClass().getSimpleName(), sal);
        }
        System.out.println("-".repeat(55));
        System.out.printf("%-20s %18s R$ %12.2f%n", "TOTAL", " ", total);
    }

    public static void main(String[] args) {
        List<Funcionario> equipe = Arrays.asList(
            new Desenvolvedor("Ana Silva", 8000, "Python"),
            new Desenvolvedor("Carlos Santos", 7500, "Java"),
            new Gerente("Mariana Costa", 12000),
            new Gerente("Roberto Lima", 11000, 2500)
        );

        System.out.println("=".repeat(60));
        System.out.println("  FOLHA DE PAGAMENTO — HERANÇA E POLIMORFISMO");
        System.out.println("=".repeat(60));

        processarFolha(equipe);
        System.out.println("\n✓ Polimorfismo: cada subclasse implementa calcularSalario() de forma diferente.");
    }
}

File: class15/python/PES1501-FuncionarioHeranca.py
"""
PES1501 - Funcionário com Herança e Polimorfismo
Aula 15: Diagrama de Classes e Orientação a Objetos
Projeto e Engenharia de Software - Prof. Giovane Barcelos

Demonstra herança, polimorfismo e encapsulamento com uma hierarquia
Funcionario → Desenvolvedor, Gerente.
"""

from abc import ABC, abstractmethod


class Funcionario(ABC):
    """Classe abstrata base — encapsula nome e salário base."""

    def __init__(self, nome, salario_base):
        self.__nome = nome
        self.__salario_base = float(salario_base)

    @property
    def nome(self):
        return self.__nome

    @property
    def salario_base(self):
        return self.__salario_base

    @abstractmethod
    def calcular_salario(self):
        """Polimorfismo: cada subclasse implementa sua regra."""
        pass

    def __repr__(self):
        return f"{self.__class__.__name__}[nome={self.__nome}, salário=R$ {self.calcular_salario():.2f}]"


class Desenvolvedor(Funcionario):
    """Desenvolvedor: salário base + 15% bônus técnico."""

    def __init__(self, nome, salario_base, linguagem):
        super().__init__(nome, salario_base)
        self.linguagem = linguagem

    def calcular_salario(self):
        return self.salario_base * 1.15


class Gerente(Funcionario):
    """Gerente: salário base + bônus fixo de R$ 2000."""

    def __init__(self, nome, salario_base, bonus=2000.0):
        super().__init__(nome, salario_base)
        self.bonus = bonus

    def calcular_salario(self):
        return self.salario_base + self.bonus


def processar_folha(funcionarios):
    """Polimorfismo: mesmo método chamado em objetos diferentes."""
    print(f"\n{'Nome':<20} {'Cargo':<18} {'Salário Final':>15}")
    print("-" * 55)
    total = 0
    for f in funcionarios:
        sal = f.calcular_salario()
        total += sal
        print(f"{f.nome:<20} {f.__class__.__name__:<18} R$ {sal:>12.2f}")
    print("-" * 55)
    print(f"{'TOTAL':<20} {' ':>18} R$ {total:>12.2f}")


if __name__ == "__main__":
    equipe = [
        Desenvolvedor("Ana Silva", 8000, "Python"),
        Desenvolvedor("Carlos Santos", 7500, "Java"),
        Gerente("Mariana Costa", 12000),
        Gerente("Roberto Lima", 11000, bonus=2500),
    ]

    print("=" * 60)
    print("  FOLHA DE PAGAMENTO — HERANÇA E POLIMORFISMO")
    print("=" * 60)

    processar_folha(equipe)
    print("\n✓ Polimorfismo: cada subclasse implementa calcular_salario() de forma diferente.")
